### ISSUE TITLE
Example single run hangs and fails to complete

### ISSUE THREAD

**Issue by rokitatomer** [t-wolfeadam/Alpyne#12]
Hello,
I am trying to run the StockManagmentGame example. Already switched various java JDK versions, but problem remains.
Here is the run output:
C:\Users\tomerro\Downloads\Stock Management Game>python "simple run - single.py"
2022-01-27 19:42:40,420 [alpyne.client.alpyne_client @ 72][    INFO] Starting model in C:\Users\tomerro\Downloads\Stock Management Game\Exported\StockManagementGame
2022-01-27 19:42:40,451 [alpyne.client.alpyne_client @ 92][   DEBUG] Executing:
java -cp ..\..\..\..\AppData\Local\Programs\Python\Python38\lib\site-packages\alpyne\resources\*;..\..\..\..\AppData\Local\Programs\Python\Python38\lib\site-packages\alpyne\resources\alpyne_lib\*;*;lib\*;lib\database\*;lib\sa\*;lib\database\querydsl\*;lib\sa\jackson\*;lib\sa\spark\* com.anylogic.alpyne.AlpyneServer -p 51150 -o C:\Users\tomerro\Downloads\Stock Management Game -l ALL .

2022-01-27 19:42:40,480 [alpyne.client.alpyne_client @ 110][    INFO] Started app | PID = 14952
2022-01-27 19:42:40,480 [alpyne.client.http_client @ 30][   DEBUG] GET /versions/number/0: None

And then the run hangs...

Any help would be appreciated.

Best,Tomer

**Comment by negodfre** [comment]
I've been struggling with the hanging as well, and no guarantee the below would work.

Today, I've been using a Jupyter-Notebook to run the simulation with less issues than running it in VSCode.
Yesterday, I ran the model several times in VSCode and then it randomly stopped working & started to hang.
After I switched to a Jupyter-Notebook, I haven't had any issues, but I made a few changes to the original code.

First, I setup a "creationflags" for the subprocess.popen within the "alpyne_client.py" file. My thought here was that when I kill a simulation or it errors out, there maybe some subprocesses left running. 

With the creation flag, I can ensure that the one subprocess is running & then I manually kill the console window after I am done.

`        cmdline_args = ["java",
                        "-cp", class_path,
                        "com.anylogic.alpyne.AlpyneServer",
                        "-p", f"{port}",
                        "-o", initdir,
                        "-l", "WARNING" if not verbose else "ALL",
                        "."]

        self.log.debug(f"Executing:\n{' '.join(cmdline_args)}\n")

        try:
            proc = subprocess.Popen(cmdline_args,
                                    stdin=subprocess.PIPE,  # Needed for quitting the app
                                    stdout=subprocess.PIPE, # Both stdout and stderr should be empty,
                                    stderr=subprocess.PIPE,
                                    creationflags=CREATE_NEW_CONSOLE) #   but open up just in case.`


Additionally, within the Jupyter-Notebook, I have a separate cell for the following:
`sim = client.create_reinforcement_learning(config)\
            .run()`

After I run this cell & I see the console window appear, I then proceed to run the simulation within the while loop.

`while not sim.is_terminal():
    status, info = sim.last_state  # updated when status is polled (automatically done thx to `blocking=True`)
    print(f"Current sim status/info = {status} | {info}")

    observation = sim.get_observation()
    ....`

Not sure if that helps at all, but I haven't found a single item that I tried that seems to resolve this issue completely.
It does seem a bit random when it stops working & starts to hang.

**Comment by rokitatomer** [comment]
Update: In order for the AlpyneClient to work, I had to add a waiting time in it's initialization (time.sleep(5) on line 46 @alpine_client.py).

**Comment by negodfre** [comment]
That worked for me as well thanks!

Oddly though, I got a new issue. 
My simulation keeps starting at the end or last model step. I haven't seen that before. 
It doesn't seem to happen if I run it from Jupyter-Notebook.

**Comment by WegmannTobias** [comment]
I've experienced the same issue, but not only with the example run but always. In my case, the problem is that AlpyneClient starts the server and asks for the version right away. If the server is not ready to respond the httpClient throws an URLError which is handled as a ModelError in the AlpyneClient and causes everything to hang. 
I've added error handling for the URLError in the httpClient. Now, if the URLError occurs, it will try to repeat the request every 10ms for 100 times. Since then I didn't have that error again. After that it would show the same behavior as before which is throwing that ModelError, but that didn't occur in my case.

```
        url_path = self.root_url + url #urllib.parse.quote(url)
        urlErrorCounter = 0
        maxUrlErrors = 100
        success = False
        while not success:
            try:
                request = Request(url_path, method=method, unverifiable=True)
                if body is not None:
                    request.data = json.dumps(body, cls=AlpyneJSONEncoder).encode('utf-8')
                request.add_header("Content-Type", "application/json")
                response = urllib.request.urlopen(request, timeout=3)
                if response.getcode() >= 400:  # TODO model errors should warn user and mark episode as halted
                    raise ModelError(response.getcode(), response.reason, response.read(), url_path)
                success = True
            except URLError as err:
                if urlErrorCounter < maxUrlErrors:
                    urlErrorCounter += 1
                    sleep(0.1)
                else:
                    # Can either be handled from alpyne or the underlying server
                    raise ModelError(err.code, err.reason, err.read(), url_path)
            except HTTPError as err:
                raise ModelError(err.code, err.reason, err.read(), url_path)
            except socket.timeout as err:
                raise ModelError(err.errno, "timeout|model exception", "see alpyne.log", url_path)
```

The other issue with the starting at the end occurs in my case if I don't have a cyclic event in my simulation which calls the takeAction-method of the RLExperiment. If you only call that on certain events or conditions and the model has no scheduled events left it shows "next event -Inf" and ends.

**Comment by t-wolfeadam** [comment]
>I've experienced the same issue, but not only with the example run but always. In my case, the problem is that AlpyneClient starts the server and asks for the version right away.
> I've added error handling for the URLError in the httpClient. Now, if the URLError occurs, it will try to repeat the request every 10ms for 100 times. Since then I didn't have that error again. 

You are correct in the reasoning; your fix is definitely more efficient than the flat 5 second sleep time mentioned above. Feel free to make a PR with these changes if you'd like and I'll be glad to approve it.

> The other issue with the starting at the end occurs in my case if I don't have a cyclic event in my simulation which calls the takeAction-method of the RLExperiment. If you only call that on certain events or conditions and the model has no scheduled events left it shows "next event -Inf" and ends.

While a separate topic, the "next event" time is pulled from the engine itself (as [documented here](https://anylogic.help/api/com/anylogic/engine/Engine.html#getnexteventtime)). Even with no scheduled events (`-infinity`), it shouldn't prematurely end your model if there's other active elements (e.g., SD stocks). I'm unable to replicate your issue - as I was able to have a model with no scheduled events that runs indefinitely - so I can't replicate what you're describing. If you can share a scenario where this does happen, please open a new issue and I'll look into it.

**Comment by t-wolfeadam** [comment]
Closing as this is likely resolved in v1.0.0. Feel free to create a new issue if it remains.
